File: Lab6-RailwayPlanning/ford_fulkerson.py
from copy import deepcopy
from find_path import find_path
from math import inf


def ford_fulkerson(G, start_index, end_index):
    """Finds the maximal flow from s to t in G"""
    # print("Entered FF")
    tot_flow = 0

    if start_index in G:  # if we've added the source node to the graph
        source_node = G[start_index]
    else:
        return 0

    if end_index in G:  # if we've added the sink node to the graph
        sink_node = G[end_index]
    else:
        return 0

    # debug_counter = 0
    path_exists = find_path(G, source_node, sink_node)  # here 'temp_point' is 't'
    while path_exists:  # while there is a path that isn't clogged up
        # print(debug_counter)
        delta = inf
        temp_node = sink_node

        while temp_node != source_node:  # find minimal delta
            delta = min(delta, temp_node.previous_edge.get_capacity(temp_node))
            temp_node = temp_node.previous_node

        tot_flow += delta
        temp_node = sink_node
        while temp_node != source_node:  # update flow along the found path
            temp_node.previous_edge.update_flow(temp_node, delta)
            temp_node = temp_node.previous_node

        path_exists = find_path(G, source_node, sink_node)

        # debug_counter += 1

    return tot_flow

File: Lab6-RailwayPlanning/C++ Solution/ford_fulkerson.cc
#include "edge.h"
#include "node.h"
#include "setup.cc"
#include <queue>
#include <limits>

// Function used in ford_fulkerson()
bool find_path(Graph& graph, Node* source, Node* sink);

/**
 * Finds the maximal flow from source_index to sink_index in graph
 */ 
int ford_fulkerson(Graph& graph, const int& source_index, const int& sink_index){
    int new_flow(0);

    Node* source = graph[source_index];
    if(source == nullptr)  // if the source node doesn't yet belong to the graph
        return 0;

    Node* sink = graph[sink_index];
    if(sink == nullptr)  // if the sink node -||-
        return 0;

    bool path_exists = find_path(graph, source, sink);
    while(path_exists){
        int delta = std::numeric_limits<int>::max();
        Node* temp_node = sink;

        // Find minimal capacity on the path found
        while(temp_node != source){
            delta = std::min(delta, temp_node->previous_edge->get_capacity(temp_node));
            temp_node = temp_node->previous_node;
        }

        // Update the flow through the found path with 'delta'
        new_flow += delta;
        temp_node = sink;
        while(temp_node != source){
            temp_node->previous_edge->update_flow(temp_node, delta);
            temp_node = temp_node->previous_node;
        }

        path_exists = find_path(graph, source, sink);
    }
    
}

/**
 * Finds a path from source to sink using BFS 
 */
bool find_path(Graph& graph, Node* source, Node* sink){

}